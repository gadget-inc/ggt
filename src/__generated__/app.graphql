"""
======================================================
THIS IS A GENERATED FILE! DO NOT EDIT IT MANUALLY!

This is copied from the gadget repository at packages/api/src/__generated__/app.graphql
We manually copy in new versions when we need to change it.
======================================================
"""

schema {
  query: Query
  mutation: Mutation
}

"Instruction for establishing a live connection that is updated once the underlying data changes."
directive @live(
  """Whether the query should be live or not."""
  if: Boolean = true
  """
  Propose a desired throttle interval ot the server in order to receive updates to at most once per "throttle" milliseconds. The server must not accept this value.
  """
  throttle: Int
) on QUERY

"""
A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the `date-time` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
"""
scalar DateTime

"""
One upload target to use for the Direct Upload style of sending files to Gadget
"""
type DirectUploadToken {
  """The token to pass to an action to reference the uploaded file."""
  token: String!
  """The URL to upload a file to."""
  url: String!
}

interface ExecutionError {
  """The Gadget platform error code for this error."""
  code: String!
  """The human facing error message for this error."""
  message: String!
  """
  The stack for any exception that caused the error. Only available for super users.
  """
  stack: String
}

type GadgetAction {
  acceptsInput: Boolean!
  apiIdentifier: String!
  availableInBulk: Boolean!
  bulkApiIdentifier: String
  bulkInvokedByIDOnly: Boolean!
  examples: GadgetActionGraphQLType
  hasAmbiguousIdentifier: Boolean!
  hasCreateOrUpdateEffect: Boolean! @deprecated(reason: "This field will be removed. Use `isCreateOrUpdateAction` instead.")
  hasDeleteRecordEffect: Boolean! @deprecated(reason: "This field will be removed. Use `isDeleteAction` instead.")
  inputFields: [GadgetObjectField!]!
  isCreateOrUpdateAction: Boolean!
  isDeleteAction: Boolean!
  name: String!
  operatesWithRecordIdentity: Boolean!
  possibleTransitions: JSONObject!
  requiresInput: Boolean!
  triggers: [GadgetTrigger!]
}

type GadgetActionGraphQLType {
  bulkOutputGraphQLTypeSDL: String
  exampleBulkGraphQLMutation: String
  exampleBulkGraphQLVariables: JSON
  exampleBulkImperativeInvocation: String
  exampleBulkMutation: String @deprecated(reason: "moved to exampleBulkGraphQLMutation")
  exampleBulkReactHook: String
  exampleGraphQLMutation: String!
  exampleGraphQLVariables: JSON!
  exampleImperativeInvocation: String!
  exampleJSInputs: JSON!
  exampleMutation: String! @deprecated(reason: "moved to exampleGraphQLMutation")
  exampleReactHook: String!
  inputGraphQLTypeSDL: String
  inputTypeScriptInterface: String
  outputGraphQLTypeSDL: String!
  outputTypeScriptInterface: String!
}

type GadgetApplicationMeta {
  allHydrations: JSON!
  """The roles that the entity making this API call has been assigned"""
  assignedRoles: [GadgetRole!]!
  canonicalRenderURL: String!
  developmentGraphQLEndpoint: String!
  developmentRenderURL: String!
  directUploadToken(
    """
    An optional unique value for this upload. Useful for forcing simultaneous requests to not be deduplicated through GraphQL client caching.
    """
    nonce: String
  ): DirectUploadToken
  editURL: String!
  environmentID: GadgetID!
  environmentName: String!
  environmentSlug: String!
  firstModelForExamples: GadgetModel!
  globalActions: [GadgetGlobalAction!]!
  graphQLEndpoint: String!
  hasGlobalActions: Boolean!
  hasLegacyEffectCards: Boolean!
  hasShopifyConnection: Boolean!
  hasSplitEnvironments: Boolean!
  hydrations(modelName: String!): HydrationPlan
  id: GadgetID!
  jsPackageIdentifier: String!
  jsPackageTarballURL: String!
  model(apiIdentifier: String, key: String): GadgetModel
  models: [GadgetModel!]!
  name: String!
  productionGraphQLEndpoint: String!
  productionRenderURL: String!
  referencedHydrations: JSON!
  roles: [GadgetRole!]!
  sessionID: String @deprecated(reason: "The current session is available as the root field `currentSession` on the root Query object, which has the ID as well as other attributes of the session.")
  shopifyConnectionApiVersion: String
  slug: String!
}

type GadgetBelongsToConfig implements GadgetFieldConfigInterface {
  fieldType: GadgetFieldType!
  isConfigured: Boolean!
  isInverseConfigured: Boolean!
  relatedModel: GadgetModel
  relatedModelKey: String
}

type GadgetDateTimeConfig implements GadgetFieldConfigInterface {
  fieldType: GadgetFieldType!
  includeTime: Boolean!
}

type GadgetEnumConfig implements GadgetFieldConfigInterface {
  allowMultiple: Boolean!
  allowOther: Boolean!
  fieldType: GadgetFieldType!
  options: [GadgetEnumOption!]!
}

type GadgetEnumOption {
  color: String!
  name: String!
}

interface GadgetField {
  apiIdentifier: String!
  configuration: GadgetFieldConfigInterface!
  fieldType: GadgetFieldType!
  hasDefault: Boolean!
  name: String!
  required: Boolean!
  requiredArgumentForInput: Boolean!
}

"""The common bits that all field configuration types share"""
interface GadgetFieldConfigInterface {
  fieldType: GadgetFieldType!
}

"""
The type of a given field of a model or other in-transit object within Gadget's type system
"""
enum GadgetFieldType {
  Any
  Array
  BelongsTo
  Boolean
  Code
  Color
  Computed
  DateTime
  Email
  EncryptedString
  Enum
  File
  HasMany
  HasManyThrough
  HasOne
  ID
  JSON
  Money
  Null
  Number
  Object
  Password
  RecordState
  RichText
  RoleAssignments
  String
  URL
  Vector
}

type GadgetFieldUsageExample {
  exampleGraphQLMutation: String!
  exampleGraphQLVariables: JSON!
  exampleImperativeInvocation: String!
  exampleReactHook: String!
}

type GadgetGenericFieldConfig implements GadgetFieldConfigInterface {
  fieldType: GadgetFieldType!
}

type GadgetGlobalAction {
  acceptsInput: Boolean!
  apiIdentifier: String!
  examples: GadgetGlobalActionGraphQLType
  name: String!
  requiresInput: Boolean!
  triggers: [GadgetTrigger!]
}

type GadgetGlobalActionGraphQLType {
  exampleGraphQLMutation: String!
  exampleGraphQLVariables: JSON!
  exampleImperativeInvocation: String!
  exampleJSInputs: JSON!
  exampleMutation: String! @deprecated(reason: "moved to exampleGraphQLMutation")
  exampleReactHook: String!
  inputGraphQLTypeSDL: String
  inputTypeScriptInterface: String
  outputGraphQLTypeSDL: String!
  outputTypeScriptInterface: String!
}

type GadgetHasManyConfig implements GadgetFieldConfigInterface {
  fieldType: GadgetFieldType!
  isConfigured: Boolean!
  isInverseConfigured: Boolean!
  relatedModel: GadgetModel
  relatedModelKey: String
}

"""The ID of a record in Gadget"""
scalar GadgetID

type GadgetModel {
  action(apiIdentifier: String!): GadgetAction
  actions: [GadgetAction!]!
  apiIdentifier: String!
  currentSingletonApiIdentifier: String
  exampleFilterQuery: String!
  exampleFindMostRecentlyCreatedQuery: String!
  exampleFullFindFirstQuery: String!
  exampleFullFindManyQuery: String!
  exampleFullFindOneQuery: String!
  exampleInternalBulkCreateMutation: String!
  exampleInternalCreateMutation: String!
  exampleInternalDeleteManyMutation: String!
  exampleInternalDeleteMutation: String!
  exampleInternalFindFirstQuery: String!
  exampleInternalFindManyQuery: String!
  exampleInternalFindOneQuery: String!
  exampleInternalUpdateMutation: String!
  examplePaginationQuery: String!
  exampleSearchQuery: String!
  exampleSimpleFindManyQuery: String!
  exampleSimpleFindOneQuery: String!
  fields: [GadgetModelField!]!
  filterGraphQLTypeName: String
  filterGraphQLTypeSDL: String
  filterable: Boolean!
  graphQLTypeName: String!
  graphQLTypeSDL: String!
  initialCreatedState: String
  key: String!
  name: String!
  pluralApiIdentifier: String!
  pluralName: String!
  searchable: Boolean!
  sortGraphQLTypeName: String
  sortGraphQLTypeSDL: String
  sortable: Boolean!
  typescriptTypeInterface: String!
  typescriptTypeInterfaceName: String!
}

"""One field of a Gadget model"""
type GadgetModelField implements GadgetField {
  apiIdentifier: String!
  configuration: GadgetFieldConfigInterface!
  examples: GadgetModelFieldExamples!
  fieldType: GadgetFieldType!
  filterable: Boolean!
  hasDefault: Boolean!
  isUniqueField: Boolean!
  name: String!
  required: Boolean!
  requiredArgumentForInput: Boolean!
  sortable: Boolean!
}

type GadgetModelFieldExamples {
  createNestedInParent: GadgetFieldUsageExample
  linkExistingChild: GadgetFieldUsageExample
  linkNewChild: GadgetFieldUsageExample
  linkToExistingParent: GadgetFieldUsageExample
}

"""One field of an action input or other transitory object in Gadget"""
type GadgetObjectField implements GadgetField {
  apiIdentifier: String!
  configuration: GadgetFieldConfigInterface!
  fieldType: GadgetFieldType!
  hasDefault: Boolean!
  name: String!
  required: Boolean!
  requiredArgumentForInput: Boolean!
}

type GadgetObjectFieldConfig implements GadgetFieldConfigInterface {
  fieldType: GadgetFieldType!
  fields: [GadgetModelField!]!
  name: String
}

"""
Represents one of the roles an identity in the system can be entitled to
"""
type GadgetRole {
  key: String!
  name: String!
  order: Int!
  selectable: Boolean!
}

type GadgetTrigger {
  specID: String!
}

"""
Instructions for a client to turn raw transport types (like strings) into useful client side types (like Dates). Unstable and not intended for developer use.
"""
scalar HydrationPlan

type InternalBulkCreateSessionsResult {
  errors: [ExecutionError!]
  sessions: [InternalSessionRecord]
  success: Boolean!
}

type InternalCreateSessionResult {
  errors: [ExecutionError!]
  session: InternalSessionRecord
  success: Boolean!
}

type InternalDeleteManySessionResult {
  errors: [ExecutionError!]
  success: Boolean!
}

type InternalDeleteSessionResult {
  errors: [ExecutionError!]
  session: InternalSessionRecord
  success: Boolean!
}

type InternalMutations {
  """Acquire a backend lock, returning only once the lock has been acquired"""
  acquireLock(
    """Lock name to acquire"""
    lock: String!
  ): LockOperationResult!
  bulkCreateSessions(sessions: [InternalSessionInput]!): InternalBulkCreateSessionsResult
  commitTransaction: String!
  createSession(session: InternalSessionInput): InternalCreateSessionResult
  deleteManySession: InternalDeleteManySessionResult
  deleteSession(id: GadgetID!): InternalDeleteSessionResult
  rollbackTransaction: String!
  startTransaction: String!
  updateSession(id: GadgetID!, session: InternalSessionInput): InternalUpdateSessionResult
}

type InternalQueries {
  """
  Currently open platform transaction details, or null if no transaction is open
  """
  currentTransactionDetails: JSONObject
  listSession(
    """Returns the items in the list that come after the specified cursor."""
    after: String
    """Returns the items in the list that come before the specified cursor."""
    before: String
    """Returns the first n items from the list."""
    first: Int
    """Returns the last n items from the list."""
    last: Int
    """
    A list of field api identifiers to return for each record. If not specified, all fields except for the computed fields on the record are returned.
    """
    select: [String!]
  ): InternalSessionRecordConnection!
  session(
    id: GadgetID!
    """
    A list of field api identifiers to return. If not specified, all fields except for the computed fields on the record are returned.
    """
    select: [String!]
  ): InternalSessionRecord
}

input InternalSessionInput {
  createdAt: DateTime
  id: GadgetID
  """A string list of Gadget platform Role keys to assign to this record"""
  roles: [String!]
  state: RecordState
  stateHistory: RecordState
  updatedAt: DateTime
}

"""
Represents one session result record in internal api calls. Returns a JSON blob of all the record's fields.
"""
scalar InternalSessionRecord

"""A connection to a list of InternalSessionRecord items."""
type InternalSessionRecordConnection {
  """A list of edges."""
  edges: [InternalSessionRecordEdge!]!
  """Information to aid in pagination."""
  pageInfo: PageInfo!
}

"""An edge in a InternalSessionRecord connection."""
type InternalSessionRecordEdge {
  """A cursor for use in pagination"""
  cursor: String!
  """The item at the end of the edge"""
  node: InternalSessionRecord!
}

type InternalUpdateSessionResult {
  errors: [ExecutionError!]
  session: InternalSessionRecord
  success: Boolean!
}

"""
This Error object represents one individual failed validation for a record field. It has a message which is appropriate for display to a user, and lists the apiIdentifier of the field in question. The `apiIdentifier` for the field is not guaranteed to exist on the model.
"""
type InvalidFieldError {
  """The apiIdentifier of the field this error occurred on."""
  apiIdentifier: String!
  """The human facing error message for this error."""
  message: String!
}

"""
This object is returned as an error when a record doesn't pass the defined validations on the model. The validation messages for each of the invalid fields are available via the other fields on this error type.
"""
type InvalidRecordError implements ExecutionError {
  """The Gadget platform error code for this InvalidRecordError."""
  code: String!
  """The human facing error message for this error."""
  message: String!
  """The model of the record which failed validation"""
  model: GadgetModel
  """
  The record which failed validation, if available. Returns all the owned fields of the record -- no sub-selections are required or possible. Only available for super users.
  """
  record: JSONObject
  """The stack for any exception that caused the error"""
  stack: String
  """
  A list of InvalidFieldError objects describing each of the errors encountered on the invalid record.
  """
  validationErrors: [InvalidFieldError!]
  """
  An object mapping field apiIdentifiers to arrays of validation error message strings for that field, as a JSON object. The object may have keys that don't correspond exactly to field apiIdentifiers if added by validations, and the object may have missing keys if no errors were encountered for that field.
  """
  validationErrorsByField: JSONObject
}

"""
The `JSON` scalar type represents JSON values as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
"""
scalar JSON

"""
The `JSONObject` scalar type represents JSON objects as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
"""
scalar JSONObject

type LockOperationResult {
  """Any errors encountered during the locking/unlocking operation"""
  errors: [ExecutionError!]
  """Whether the lock operation succeeded"""
  success: Boolean!
}

type Mutation {
  """
  Meta information about the application, like it's name, schema, and other internal details.
  """
  gadgetMeta: GadgetApplicationMeta!
  internal: InternalMutations
}

"""Information about pagination in a connection."""
type PageInfo {
  """When paginating forwards, the cursor to continue."""
  endCursor: String
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!
  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!
  """When paginating backwards, the cursor to continue."""
  startCursor: String
}

type Query {
  currentSession: Session
  """
  Meta information about the application, like it's name, schema, and other internal details.
  """
  gadgetMeta: GadgetApplicationMeta!
  internal: InternalQueries
  session(id: GadgetID!): Session
  sessions(
    """Returns the items in the list that come after the specified cursor."""
    after: String
    """Returns the items in the list that come before the specified cursor."""
    before: String
    """Returns the first n items from the list."""
    first: Int
    """Returns the last n items from the list."""
    last: Int
  ): SessionConnection!
}

"""
Represents the state of one record in a Gadget database. Represented as either a string or set of strings nested in objects.
"""
scalar RecordState

"""
A named group of permissions granted to a particular actor in the system. Managed in the Gadget editor.
"""
type Role {
  """
  The stable identifier for this role. Null if the role has since been deleted.
  """
  key: String!
  """The human readable name for this role. Can be changed."""
  name: String!
}

type Session {
  """
  Get all the fields for this record. Useful for not having to list out all the fields you want to retrieve, but slower.
  """
  _all: JSONObject!
  """
  The time at which this record was first created. Set once upon record creation and never changed. Managed by Gadget.
  """
  createdAt: DateTime!
  """
  The globally unique, unchanging identifier for this record. Assigned and managed by Gadget.
  """
  id: GadgetID
  roles: [Role!]
  """
  The time at which this record was last changed. Set each time the record is successfully acted upon by an action. Managed by Gadget.
  """
  updatedAt: DateTime!
}

"""A connection to a list of Session items."""
type SessionConnection {
  """A list of edges."""
  edges: [SessionEdge!]!
  """Information to aid in pagination."""
  pageInfo: PageInfo!
}

"""An edge in a Session connection."""
type SessionEdge {
  """A cursor for use in pagination"""
  cursor: String!
  """The item at the end of the edge"""
  node: Session!
}

"""
This Error object is returned for errors which don't have other specific handling. It has a message which is safe to display to users, but is often technical in nature. It also has a `code` field which is documented in the Gadget API Error Codes docs.
"""
type SimpleError implements ExecutionError {
  """The Gadget platform error code for this error."""
  code: String!
  """The human facing error message for this error."""
  message: String!
  """The stack for any exception that caused the error"""
  stack: String
}