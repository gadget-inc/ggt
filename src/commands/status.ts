import { ArgError } from "../services/command/arg.js";
import type { Command, Usage } from "../services/command/command.js";
import { getConflicts, printConflicts } from "../services/filesync/conflicts.js";
import { UnknownDirectoryError } from "../services/filesync/error.js";
import { FileSync } from "../services/filesync/filesync.js";
import { SyncJson, SyncJsonArgs, loadSyncJsonDirectory } from "../services/filesync/sync-json.js";
import { sprint } from "../services/output/sprint.js";

export type StatusArgs = typeof args;

export const args = SyncJsonArgs;

export const usage: Usage = () => {
  return sprint`
    Shows file changes since last sync (e.g. $ggt dev, push, deploy etc.)

    {gray Usage}
          ggt status
  `;
};

export const command: Command<StatusArgs> = async (ctx) => {
  if (ctx.args._.length > 0) {
    throw new ArgError(sprint`
      "ggt status" does not take any positional arguments.

      If you are trying to see the status of a specific directory,
      you must "cd" to that directory and then run "ggt status".

      Run "ggt status -h" for more information.
    `);
  }

  const directory = await loadSyncJsonDirectory(process.cwd());
  const syncJson = await SyncJson.load(ctx, { directory });
  if (!syncJson) {
    throw new UnknownDirectoryError(ctx, { directory });
  }

  syncJson.print();

  const filesync = new FileSync(syncJson);
  const hashes = await filesync.hashes(ctx);
  await filesync.print(ctx, { hashes });

  const conflicts = getConflicts({ localChanges: hashes.localChanges, environmentChanges: hashes.environmentChanges });
  if (conflicts.size > 0) {
    ctx.log.debug("conflicts detected", { conflicts });
    printConflicts({ conflicts });
  }
};
