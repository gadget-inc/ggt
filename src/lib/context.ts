import type { Config } from "@oclif/core";
import assert from "assert";
import fs from "fs-extra";
import got, { HTTPError } from "got";
import { isString } from "lodash";
import path from "path";
import { Env } from "./env";
import { ignoreEnoent } from "./fs-utils";

export const GADGET_ENDPOINT = Env.productionLike ? "https://app.gadget.dev" : "https://app.ggt.dev:3000";

class Context {
  /**
   * A reference to oclif's {@linkcode Config}.
   *
   * By default, oclif's {@linkcode Config} is only available as an instance property on a Command, but we want to be able to access it from
   * anywhere. To do this, we created this global variable that references the Config. It is set by the init function in the BaseCommand.
   */
  config!: Config;

  app?: App;

  private _session?: string;

  private _user?: User;

  private _availableApps: App[] = [];

  private _request = got.extend({
    hooks: {
      beforeRequest: [
        (options) => {
          options.headers["user-agent"] = this.config.userAgent;
          if (options.url.origin === GADGET_ENDPOINT && this.session) {
            options.headers["cookie"] = `session=${encodeURIComponent(this.session)};`;
          }
        },
      ],
    },
  });

  get session(): string | undefined {
    if (this._session) return this._session;

    try {
      this._session = fs.readFileSync(path.join(this.config.configDir, "session.txt"), "utf-8");
      return this._session;
    } catch (error) {
      ignoreEnoent(error);
      return undefined;
    }
  }

  set session(value: string | undefined) {
    this.clear();
    this._session = value;
    if (this._session) {
      fs.outputFileSync(path.join(this.config.configDir, "session.txt"), this._session);
    } else {
      fs.removeSync(path.join(this.config.configDir, "session.txt"));
    }
  }

  /**
   * @returns The current user, or undefined if the user is not logged in.
   */
  async getUser(): Promise<User | undefined> {
    if (!this.session) return undefined;
    if (this._user) return this._user;

    try {
      this._user = await this._request(`${GADGET_ENDPOINT}/auth/api/current-user`).json<User>();
      return this._user;
    } catch (error) {
      if (error instanceof HTTPError && error.response.statusCode === 401) {
        this.session = undefined;
        return undefined;
      }
      throw error;
    }
  }

  /**
   * @returns The list of Gadget applications the current user has access to.
   */
  async getAvailableApps(): Promise<App[]> {
    if (!this.session) return [];
    if (this._availableApps.length > 0) return this._availableApps;

    this._availableApps = await this._request(`${GADGET_ENDPOINT}/auth/api/apps`).json<App[]>();
    return this._availableApps;
  }

  async setApp(value: App | string | undefined): Promise<void> {
    if (isString(value)) {
      const app = await this.getAvailableApps().then((apps) => apps.find((app) => app.name == value || app.slug == value));
      assert(app, `attempted to set app to "${value}" but no app with that name or slug was found`);
      this.app = app;
    } else {
      this.app = value;
    }
  }

  clear(): void {
    this._session = undefined;
    this._user = undefined;
    this.app = undefined;
    this._availableApps = [];
  }
}

export const context = new Context();

export interface User {
  email: string;
  name?: string;
}

export interface App {
  id: string | number;
  name: string;
  slug: string;
}
