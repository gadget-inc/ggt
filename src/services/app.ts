import { map } from "lodash";
import { z } from "zod";
import { config } from "./config.js";
import { http, loadCookie } from "./http.js";
import type { User } from "./user.js";

export const App = z.object({
  id: z.union([z.string(), z.number(), z.bigint()]),
  slug: z.string(),
  primaryDomain: z.string(),
  hasSplitEnvironments: z.boolean(),
});

export type App = z.infer<typeof App> & { user: User };

/**
 * @returns The list of Gadget applications the current user has access to.
 */
export const getApps = async (user: User): Promise<App[]> => {
  const cookie = loadCookie();
  if (!cookie) {
    return [];
  }

  const json = await http({
    url: `https://${config.domains.services}/auth/api/apps`,
    headers: { cookie },
    responseType: "json",
    resolveBodyOnly: true,
  });

  return map(z.array(App).parse(json), (app) => ({ ...app, user }));
};
