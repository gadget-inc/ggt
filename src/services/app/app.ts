import assert from "node:assert";
import { z } from "zod";
import type { Context } from "../command/context.js";
import { config } from "../config/config.js";
import { loadCookie } from "../http/auth.js";
import { http } from "../http/http.js";

export const EnvironmentType = Object.freeze({
  Development: "development",
  Production: "production",
  Test: "test",
});

export type EnvironmentType = keyof typeof EnvironmentType;

export const Environment = z.object({
  id: z.union([z.string(), z.number(), z.bigint()]).transform((v) => BigInt(v)),
  name: z.string(),
  type: z.nativeEnum(EnvironmentType),
});

export type Environment = z.infer<typeof Environment>;

// TODO: rename to Application
export const App = z.object({
  id: z.union([z.string(), z.number(), z.bigint()]).transform((v) => BigInt(v)),
  slug: z.string(),
  primaryDomain: z.string(),
  hasSplitEnvironments: z.boolean(),
  multiEnvironmentEnabled: z.boolean(),
  environments: z.array(Environment),
});

export type App = z.infer<typeof App>;

/**
 * Retrieves a list of apps for the given user. If the user is not
 * logged in, an empty array is returned instead.
 *
 * @param ctx - The current context.
 * @returns A promise that resolves to an array of App objects.
 */
// TODO: cache this
export const getApps = async (ctx: Context): Promise<App[]> => {
  const cookie = loadCookie();
  if (!cookie) {
    return [];
  }

  assert(ctx.user, "must get user before getting apps");

  const json = await http({
    context: { ctx },
    url: `https://${config.domains.services}/auth/api/apps`,
    headers: { cookie },
    responseType: "json",
    resolveBodyOnly: true,
  });

  return z.array(App).parse(json);
};
