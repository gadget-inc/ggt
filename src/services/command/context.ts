import { createLogger, type Logger } from "../output/log/logger.js";
import type { StructuredLoggerOptions } from "../output/log/structured.js";
import { PromiseSignal } from "../util/promise.js";
import type { AnyVoid } from "../util/types.js";

/**
 * Represents the context of a command-line operation.
 */
export class Context extends AbortController {
  /**
   * A promise that resolves when the context is aborted and all the
   * registered onAbort callbacks have finished.
   */
  readonly done = new PromiseSignal<void>();

  /**
   * The logger for the current context.
   */
  #log: Logger;

  /**
   * The values that have been set on this context.
   */
  #values: Record<symbol, unknown>;

  /**
   * The callbacks that will be called when this context is aborted.
   */
  #onAborts: OnAbort[] = [];

  private constructor({ log, values }: { log: Logger; values: Record<symbol, unknown> }) {
    super();
    this.#log = log;
    this.#values = values;

    // in case this context is ...spread into another object
    this.abort = this.abort.bind(this);
    this.child = this.child.bind(this);
    this.onAbort = this.onAbort.bind(this);

    // when the context is aborted, call all the registered callbacks
    this.signal.addEventListener(
      "abort",
      // eslint-disable-next-line @typescript-eslint/no-misused-promises
      async () => {
        let error: unknown;

        // call the callbacks in reverse order, like go's defer
        for (const callback of this.#onAborts.reverse()) {
          try {
            await callback(this.signal.reason);
          } catch (e: unknown) {
            error ??= e;
            this.log.error("error during abort", { error });
          }
        }

        if (error) {
          this.done.reject(error);
        } else {
          this.done.resolve();
        }
      },
    );
  }

  /**
   * A {@linkcode Logger} that can print to stdout and log structured
   * messages to stderr.
   */
  get log(): Logger {
    return this.#log;
  }

  /**
   * Initializes a new context.
   */
  static init(options: StructuredLoggerOptions): Context {
    return new Context({
      log: createLogger(options),
      values: {},
    });
  }

  get(key: symbol): unknown {
    return this.#values[key];
  }

  set(key: symbol, value: unknown): void {
    this.#values[key] = value;
  }

  /**
   * Returns a new context that is a child of the current context.
   *
   * @see {@linkcode ContextInit}
   */
  child(options: StructuredLoggerOptions): Context {
    const ctx = new Context({
      log: this.log.child(options),
      values: this.#values,
    });

    this.onAbort(() => ctx.abort());

    return ctx;
  }

  /**
   * Registers a callback that will be called when the context is
   * aborted (e.g. when the user presses Ctrl+C).
   *
   * @param callback - The callback to call when the context is aborted.
   */
  onAbort(callback: OnAbort): void {
    this.#onAborts.push(callback);
  }
}

/**
 * A callback that will be called when the context is aborted.
 */
export type OnAbort = (reason: unknown) => AnyVoid;
