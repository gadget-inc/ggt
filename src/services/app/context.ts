import assert from "node:assert";
import { type Context } from "../command/context.js";
import type { Application, Environment } from "./app.js";

const kApp = Symbol.for("app");
const kEnv = Symbol.for("env");

export const maybeGetCurrentApp = (ctx: Context): Application | undefined => {
  return ctx.get(kApp) as Application | undefined;
};

export const getCurrentApp = (ctx: Context): Application => {
  const app = maybeGetCurrentApp(ctx);
  assert(app, "missing app in context");
  return app;
};

export const setCurrentApp = (ctx: Context, app: Application): void => {
  ctx.set(kApp, app);
};

export const maybeGetCurrentEnv = (ctx: Context): Environment | undefined => {
  return ctx.get(kEnv) as Environment | undefined;
};

export const getCurrentEnv = (ctx: Context): Environment => {
  const env = maybeGetCurrentEnv(ctx);
  assert(env, "missing env in context");
  return env;
};

export const setCurrentEnv = (ctx: Context, env: Environment): void => {
  ctx.set(kEnv, env);
};
