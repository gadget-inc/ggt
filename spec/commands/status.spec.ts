import { beforeEach, describe, it } from "vitest";
import { args, command as status, type StatusArgs } from "../../src/commands/status.js";
import type { Context } from "../../src/services/command/context.js";
import { nockTestApps } from "../__support__/app.js";
import { makeContext } from "../__support__/context.js";
import { makeSyncScenario } from "../__support__/filesync.js";
import { expectStdout } from "../__support__/output.js";
import { loginTestUser } from "../__support__/user.js";

describe("status", () => {
  let ctx: Context<StatusArgs>;

  beforeEach(() => {
    loginTestUser();
    nockTestApps();

    ctx = makeContext({
      parse: args,
      argv: ["status"],
    });
  });

  it("prints the expected message when nothing has changed", async () => {
    await makeSyncScenario({
      localFiles: {
        ".gadget/": "",
      },
    });

    await status(ctx);

    expectStdout().toMatchSnapshot();
  });

  it("prints the expected message when local files have changed", async () => {
    await makeSyncScenario({
      localFiles: {
        ".gadget/": "",
        "local-file.txt": "changed",
      },
    });

    await status(ctx);

    expectStdout().toMatchSnapshot();
  });

  it("prints the expected message when gadget files have changed", async () => {
    await makeSyncScenario({
      localFiles: {
        ".gadget/": "",
      },
      gadgetFiles: {
        "gadget-file.txt": "changed",
      },
    });

    await status(ctx);

    expectStdout().toMatchSnapshot();
  });

  it("prints the expected message when both local and gadget files have changed", async () => {
    await makeSyncScenario({
      localFiles: {
        ".gadget/": "",
        "local-file.txt": "changed",
      },
      gadgetFiles: {
        "gadget-file.txt": "changed",
      },
    });

    await status(ctx);

    expectStdout().toMatchSnapshot();
  });
});
