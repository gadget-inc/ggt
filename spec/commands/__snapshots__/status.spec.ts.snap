// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`status > prints the expected message when both local and gadget files have changed 1`] = `
"Application  test
Environment  development
     Branch  test-branch

Your local filesystem has changed.
+  local-file.txt  created 

Your environment's filesystem has changed.
+  gadget-file.txt  created 
"
`;

exports[`status > prints the expected message when gadget files have changed 1`] = `
"Application  test
Environment  development
     Branch  test-branch

Your local filesystem has not changed.

Your environment's filesystem has changed.
+  gadget-file.txt  created 
"
`;

exports[`status > prints the expected message when local files have changed 1`] = `
"Application  test
Environment  development
     Branch  test-branch

Your local filesystem has changed.
+  local-file.txt  created 

Your environment's filesystem has not changed.
"
`;

exports[`status > prints the expected message when nothing has changed 1`] = `
"Application  test
Environment  development
     Branch  test-branch

Your local filesystem has not changed.

Your environment's filesystem has not changed.
"
`;
